cpp/server/proxy.cc

```diff
@@ -0,0 +1,134 @@
+  for (auto it(response->headers.begin()); it != response->headers.end();
+       ++it) {
+    CHECK_EQ(evhttp_add_header(evhttp_request_get_output_headers(request),
+                               it->first.c_str(), it->second.c_str()),
+             0);
+  }
+  CHECK_GT(evbuffer_add_printf(evhttp_request_get_output_buffer(request), "%s",
+                               response->body.c_str()),
+           -1);
+
+  const int response_code(response->status_code);
+  base->Add([request, response_code]() {
+    evhttp_send_reply(request, response_code, /*reason*/ NULL,
+                      /*databuf*/ NULL);
+  });
+}
+
+
+}  // namespace
+
+
+Proxy::Proxy(libevent::Base* base, JsonOutput* output,
+             const GetFreshNodesFunction& get_fresh_nodes, UrlFetcher* fetcher,
+             util::Executor* executor)
+    : base_(base),
+      output_(output),
+      get_fresh_nodes_(get_fresh_nodes),
+      fetcher_(fetcher),
+      executor_(executor) {
+}
+
+
+void Proxy::ProxyRequest(evhttp_request* req) const {
+  CHECK_NOTNULL(req);
+
+  const std::vector<ct::ClusterNodeState> fresh_nodes(get_fresh_nodes_());
+  if (fresh_nodes.empty()) {
+    return output_->SendError(req, HTTP_SERVUNAVAIL,
+                              "No node able to serve request.");
+  }
+  const ct::ClusterNodeState& target(
+      fresh_nodes[std::rand() % fresh_nodes.size()]);
+
+  URL url(evhttp_request_uri(req));
+  url.SetProtocol("http");
+  url.SetHost(target.hostname());
+  url.SetPort(target.log_port());
+
+  UrlFetcher::Request fetcher_req(url);
+
+  switch (evhttp_request_get_command(req)) {
+    case EVHTTP_REQ_DELETE:
+      fetcher_req.verb = UrlFetcher::Verb::DELETE;
+      break;
+    case EVHTTP_REQ_GET:
+      fetcher_req.verb = UrlFetcher::Verb::GET;
+      break;
+    case EVHTTP_REQ_POST:
+      fetcher_req.verb = UrlFetcher::Verb::POST;
+      break;
+    case EVHTTP_REQ_PUT:
+      fetcher_req.verb = UrlFetcher::Verb::PUT;
+      break;
+    default:
+      return output_->SendError(req, HTTP_BADMETHOD, "Bad method requested.");
+      break;
+  }
+
+  for (evkeyval* ptr = evhttp_request_get_input_headers(req)->tqh_first; ptr;
+       ptr = ptr->next.tqe_next) {
+    fetcher_req.headers.insert(std::make_pair(ptr->key, ptr->value));
+  }
+  if (fetcher_req.verb == UrlFetcher::Verb::PUT ||
+      fetcher_req.verb == UrlFetcher::Verb::POST) {
+    const size_t body_length(
+        evbuffer_get_length(evhttp_request_get_input_buffer(req)));
+    std::string body(reinterpret_cast<const char*>(evbuffer_pullup(
+                         evhttp_request_get_input_buffer(req), body_length)),
+                     body_length);
+    fetcher_req.body.swap(body);
+  }
+  VLOG(1) << "Proxying request to " << url.Host() << ":" << url.Port()
+          << url.PathQuery();
+  UrlFetcher::Response* resp(new UrlFetcher::Response);
+  fetcher_->Fetch(fetcher_req, resp,
+                  new util::Task(std::bind(&ProxyRequestDone, base_, output_,
```

Comment:
Yeah, a hangover from it originally being a template.
added a bunch of `using`s.